Route all tests through d_do_test

marler8997 opened the pull request:
Provides a common entry point for all tests to perform common setup/operations.  Also simplifies the interface to run a test (resulting in a simplified Makefile).

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -532,20 +528,35 @@ int main(string[] args)
-    string test_name      = args[2];
-    string test_extension = args[3];
+    auto test_file = args[1];
+    auto result = matchFirst(test_file, `^([^/^\\]+)[/\\]([^\.]+)\.(.+)$`);
```
Does this need to be RegExp? Doesn't baseName + stripExtension work too?

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -532,20 +528,35 @@ int main(string[] args)
+    {
+        auto process = spawnProcess(["./tools/sh_do_test.sh".replace("/", envData.sep),
+            input_dir, test_name]);
+        return process.wait();
```
Why not `std.process.execute` if you wait anyhow?

marler8997 replied:
not capturing stdout/stderr

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -532,20 +529,29 @@ int main(string[] args)
+    auto test_file = args[1];
+    string input_dir = test_file.dirName;
+    string test_base_name = test_file.baseName;
+    string test_name = test_base_name.stripExtension();
```
This is a bit inconsistent here. All three are functions ;-)
(by "your majority vote" maybe adapt stripExtension?)

wilzbach commented on test/tools/d_do_test.d:
```diff
@@ -532,20 +529,29 @@ int main(string[] args)
+    if (test_base_name.extension() == ".sh")
+    {
+        auto process = spawnShell(format("%s %s %s",
+            "./tools/sh_do_test.sh".replace("/", envData.sep), input_dir, test_name));
```
`"tools".buildPath("sh_do_test.sh")` ?

wilzbach reviewed (approved):
Nice & one step closer to making it easier for people on Windows to run the testsuite :)

marler8997 committed:
Route all tests through d_do_test

Provides a common entry point for all tests to perform common setup/operations.  Also simplifies the interface to run a test.

wilzbach commented:
```
C:\projects\dmd\test\tools\sh_do_test.sh: line 17: /c/projects/dmd/test/exported_vars.sh: No such file or directory
C:\projects\dmd\test\tools\sh_do_test.sh: line 17: /c/projects/dmd/test/exported_vars.sh: No such file or directory
make[1]: *** [Makefile:129: test_results/runnable/test35.sh.out] Error 1
make[1]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:129: test_results/runnable/test44.sh.out] Error 1
 ... runnable\ifti.d                  (-O -inline -g)
make: *** [Makefile:152: start_runnable_tests] Error 2
```

ibuclaw commented:
Can't merge if there's failing tests, so that will need to be addressed first. Other than that, so far seems reasonable.

wilzbach commented:
FYI I already restarted it twice before posting the error, so I think the error is a permanent one. It probably has something to do with AppVeyor's setup of Cygwin.

wilzbach commented on test/tools/sh_do_test.sh:
```diff
@@ -8,13 +8,11 @@ if [ "${RESULTS_DIR}" == "" ]; then
     exit 1
 fi
 
-DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
Why did you remove this?
This allows the script to be called from anywhere (in the future at least)

marler8997 replied:
I think this might be why AppVeyor is failing...I"ve removed it to see if it fixes the issue... I know removing it is more "brittle".  If this works, maybe I should redefined it to something like `${TEST_DIR}/../tools`?  I don't like it as much.  The fact that Bash script files don't have an easy way to get their own directory is such a pain.  This is one thing that BATCH has over Bash :) (`%~dp0`)

marler8997 replied:
Adding `bash` to the beginning of the `spawnShell` command seems to have fixed the AppVeyor issue.

marler8997 commented:
Yeah, I've been seeing a bunch of different errors on AppVeyor recently in other PRs but this one is consistent.  Looks like it might be the definition of `DIR` in `sh_do_test.sh` might not be working for the AppVeyor...on to trial/error
